source-postgres/discovery.go

```diff
@@ -68,20 +72,58 @@ func (db *postgresDatabase) DiscoverTables(ctx context.Context) (map[string]*sql
 			tableMap[streamID] = info
 		}
 	}
+
 	for streamID, key := range primaryKeys {
 		// The `getColumns()` query implements the "exclude system schemas" logic,
 		// so here we ignore primary key information for tables we don't care about.
 		var info, ok = tableMap[streamID]
 		if !ok {
 			continue
 		}
-		logrus.WithFields(logrus.Fields{"table": streamID, "key": key}).Trace("queried primary key")
+		logrus.WithFields(logrus.Fields{
+			"table": streamID,
+			"key":   key,
+		}).Trace("queried primary key")
 		info.PrimaryKey = key
 		tableMap[streamID] = info
 	}
+
+	// For tables which have no primary key but have a valid secondary index,
+	// fill in that secondary index as the 'primary key' for our purposes.
+	for streamID, indexColumns := range secondaryIndexes {
+		var info, ok = tableMap[streamID]
+		if !ok || info.PrimaryKey != nil {
```

Comment:
Exactly, if there are multiple indexes which satisfy all of our requirements we'll just arbitrarily pick one of them (at random, due to how Go map iteration order works).

This does mean that re-running discovery could potentially suggest a different key each time, but we previously decided that the catalog merge logic should always preserve the original key for preexisting collections upon rediscovery (though I don't know if that's been changed yet, so it could misbehave in practice until the logic gets fixed there)